src/Transports.AspNetCore/WebSockets/SubscriptionsTransportWs/SubscriptionServer.cs

```diff
@@ -0,0 +1,250 @@
+            else
+            {
+                await OnConnectionInitAsync(message, false);
+            }
+            return;
+        }
+        if (!Initialized)
+        {
+            await ErrorNotInitializedAsync(message);
+            return;
+        }
+        switch (message.Type)
+        {
+            case MessageType.GQL_START:
+                await OnStartAsync(message);
+                break;
+            case MessageType.GQL_STOP:
+                await OnStopAsync(message);
+                break;
+            default:
+                await ErrorUnrecognizedMessageAsync(message);
+                break;
+        }
+    }
+
+    private static readonly OperationMessage _keepAliveMessage = new() { Type = MessageType.GQL_CONNECTION_KEEP_ALIVE };
+    /// <inheritdoc/>
+    protected override Task OnSendKeepAliveAsync()
+        => Connection.SendMessageAsync(_keepAliveMessage);
+
+    private static readonly OperationMessage _connectionAckMessage = new() { Type = MessageType.GQL_CONNECTION_ACK };
+    /// <inheritdoc/>
+    protected override Task OnConnectionAcknowledgeAsync(OperationMessage message)
+        => Connection.SendMessageAsync(_connectionAckMessage);
+
+    /// <summary>
+    /// Executes when a request is received to start a subscription.
+    /// </summary>
+    protected virtual Task OnStartAsync(OperationMessage message)
+        => SubscribeAsync(message, true);
+
+    /// <summary>
+    /// Executes when a request is received to stop a subscription.
+    /// </summary>
+    protected virtual Task OnStopAsync(OperationMessage message)
+        => UnsubscribeAsync(message.Id);
+
+    /// <inheritdoc/>
+    protected override async Task SendErrorResultAsync(string id, ExecutionResult result)
+    {
+        if (Subscriptions.TryRemove(id))
+        {
+            await Connection.SendMessageAsync(new OperationMessage
+            {
+                Id = id,
+                Type = MessageType.GQL_ERROR,
+                Payload = result,
+            });
+        }
+    }
+
+    /// <inheritdoc/>
+    protected override async Task SendDataAsync(string id, ExecutionResult result)
+    {
+        if (Subscriptions.Contains(id))
+        {
+            await Connection.SendMessageAsync(new OperationMessage
+            {
+                Id = id,
+                Type = MessageType.GQL_DATA,
+                Payload = result,
+            });
+        }
+    }
+
+    /// <inheritdoc/>
+    protected override async Task SendCompletedAsync(string id)
+    {
+        if (Subscriptions.TryRemove(id))
+        {
+            await Connection.SendMessageAsync(new OperationMessage
+            {
+                Id = id,
+                Type = MessageType.GQL_COMPLETE,
+            });
+        }
+    }
+
+    /// <inheritdoc/>
+    protected override async Task<ExecutionResult> ExecuteRequestAsync(OperationMessage message)
+    {
+        var request = Serializer.ReadNode<GraphQLRequest>(message.Payload)!;
```

Comment:
Payload may be `null` here, better to check.

Reply:
Added check which throws `ArgumentNullException` here, which is caught by `SubscribeAsync`, encapsulated into an `ExecutionError` by `HandleErrorDuringSubscribeAsync`, and then sent as a connection-attempt-failure and removed from the subscription list by `SendErrorResultAsync`.  This is similar to what would happen if a deserialization error were to occur.